Userland/Utilities/headless-browser.cpp

```diff
@@ -269,49 +316,40 @@ static ErrorOr<NonnullRefPtr<Core::Timer>> load_page_for_screenshot_and_exit(Cor
     return timer;
 }
 
-static ErrorOr<TestResult> run_dump_test(HeadlessWebContentView& view, URL::URL const& url, StringView expectation_path, TestMode mode, int timeout_in_milliseconds = DEFAULT_TIMEOUT_MS)
+static void run_dump_test(HeadlessWebContentView& view, Test& test, URL::URL const& url, int timeout_in_milliseconds = DEFAULT_TIMEOUT_MS)
 {
-    Core::EventLoop loop;
-    bool did_timeout = false;
+    auto timer = Core::Timer::create_single_shot(timeout_in_milliseconds, [&view, &test]() {
+        view.on_load_finish = {};
+        view.on_text_test_finish = {};
 
-    auto timeout_timer = Core::Timer::create_single_shot(timeout_in_milliseconds, [&] {
-        did_timeout = true;
-        loop.quit(0);
+        view.on_test_complete({ test, TestResult::Timeout });
```

Comment:
it feels kind of sad to "resolve" the promise with a timeout, rather than reject it.

What happens if one of the try's in the test methods fail to occur?

.. I guess the real comment is that I see a lot of "when_resolved" and not a lot of "when_rejected", which sounds like it could lead to some hard-to-debug timeouts?

Reply:
The TRYs are converted to `TestResult::Error` in `on_test_complete`.

The main reason I didn't bother with `when_rejected` is because we want to do the same thing no matter what happens to the test - track the failed test for logging, and move on to the next one. Note that the only handler of this particular promise is the one in `run_tests`. Conditionally rejecting or resolving that promise would just mean duplicating that handler.